test/functional/legacy/cdo_spec.lua

```diff
@@ -0,0 +1,198 @@
+      :  " If the executed command fails, then the operation should be aborted
+      :  enew!
+      :  let subst_count = 0
+      :  exe a:cchar . "do s/Line/xLine/ | let subst_count += 1"
+      :  if subst_count == 1 && getline('.') == 'xLine1'
+      :     let g:result .= 'Abort command on error test passed' . nl
+      :  else
+      :     let g:result .= 'Abort command on error test failed' . nl
+      :  endif
+
+      :  exe "2,2" . a:cchar . "do! let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+
+      :  " List with no valid error entries
+      :  edit! +2 Xtestfile1
+      :  exe a:cchar . "getexpr ['non-error 1', 'non-error 2', 'non-error 3']"
+      :  exe a:cchar . "do let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+      :  exe "2" . a:cchar . "do let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+      :  let v:errmsg=''
+      :  exe "%" . a:cchar . "do let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+      :  exe "1,$" . a:cchar . "do let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+      :  exe "." . a:cchar . "do let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+      :  let g:result .= v:errmsg
+
+      :  " List with only one valid entry
+      :  exe a:cchar . "getexpr ['Xtestfile3:3:1:Line3']"
+      :  exe a:cchar . "do let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+
+      :  " Tests for :cfdo and :lfdo commands
+      :  exe a:cchar . "getexpr ['non-error 1', 'Xtestfile1:1:3:Line1', 'Xtestfile1:2:1:Line2', 'non-error 2', 'Xtestfile2:2:2:Line2', 'non-error 3', 'Xtestfile3:2:3:Line2', 'Xtestfile3:3:1:Line3']"
+      :  exe a:cchar . "fdo let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+      :  exe "3" . a:cchar . "fdo let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+      :  exe "2,3" . a:cchar . "fdo let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+      :  exe "%" . a:cchar . "fdo let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+      :  exe "1,$" . a:cchar . "fdo let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+      :  exe a:cchar . 'pfile'
+      :  exe "." . a:cchar . "fdo let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+
+      :  " List with only one valid entry
+      :  exe a:cchar . "getexpr ['Xtestfile2:2:5:Line2']"
+      :  exe a:cchar . "fdo let g:result .= expand('%') . ' ' . line('.') . 'L' . ' ' . col('.') . 'C' . nl"
+      :endfunction
+    ]=])
+  end)
+
+  after_each(function()
+    os.remove('Xtestfile1')
+    os.remove('Xtestfile2')
+    os.remove('Xtestfile3')
+  end)
+
+  it('works for :cdo', function()
+    execute("let result=''")
+
+    feed(":call RunTests('c')<c-l>")
```

Comment:
This also avoids the hang:

    execute([[call RunTests('c')]])
    helpers.run(nil, nil, nil, 100)